encoding/cbf/common_codec/common.go

```diff
@@ -0,0 +1,505 @@
+		boolean = true
+	default:
+		err = CodecError(fmt.Sprintf("invalid boolean value: %d", b[0]))
+	}
+
+	return
+}
+
+//
+// Length
+//
+
+// TODO encode length with variable-sized encoding?
+//      e.g. first byte starting with `0` is the last byte in the length
+//      will usually save 3 bytes. the question is if it saves or costs encode and/or decode time
+
+// EncodeLength encodes a non-negative length as a uint32.
+// It uses 4 bytes.
+func EncodeLength(w io.Writer, length int) (err error) {
+	if length < 0 { // TODO is this safety check useful?
+		return CodecError(fmt.Sprintf("cannot encode length below zero: %d", length))
+	}
+
+	l := uint32(length)
+
+	return binary.Write(w, binary.BigEndian, l)
+}
+
+func DecodeLength(r io.Reader) (length int, err error) {
+	b := make([]byte, 4)
+
+	bytesRead, err := r.Read(b)
+	if err != nil {
+		return
+	}
+	if bytesRead != 4 {
+		err = CodecError("EOF when reading length")
+		return
+	}
+
+	asUint32 := binary.BigEndian.Uint32(b)
+	length = int(asUint32)
+	return
+}
+
+//
+// Bytes
+//
+
+func EncodeBytes(w io.Writer, bytes []byte) (err error) {
+	err = EncodeLength(w, len(bytes))
+	if err != nil {
+		return
+	}
+	_, err = w.Write(bytes)
+	return
+}
+
+func DecodeBytes(r io.Reader) (bytes []byte, err error) {
+	length, err := DecodeLength(r)
+	if err != nil {
+		return
+	}
+
+	bytes = make([]byte, length)
+
+	bytesRead, err := r.Read(bytes)
+	if err != nil {
+		return
+	}
+	if bytesRead != length {
+		err = CodecError("EOF when reading bytes")
+		return
+	}
+
+	return
+}
+
+//
+// String
+//
+
+func EncodeString(w io.Writer, s string) (err error) {
+	return EncodeBytes(w, []byte(s))
+}
+
+func DecodeString(r io.Reader) (s string, err error) {
+	b, err := DecodeBytes(r)
+	if err != nil {
+		return
+	}
+	s = string(b)
+	return
+}
+
+//
+// Address
+//
+
+func EncodeAddress[Address common.Address | cadence.Address](w io.Writer, a Address) (err error) {
+	_, err = w.Write(a[:])
+	return
+}
+
+func DecodeAddress(r io.Reader) (a common.Address, err error) {
+	bytes := make([]byte, common.AddressLength)
+
+	bytesRead, err := r.Read(bytes)
+	if err != nil {
+		return
+	}
+	if bytesRead != common.AddressLength {
+		err = CodecError("EOF when reading address")
+		return
+	}
+
+	return common.BytesToAddress(bytes)
+}
+
+//
+// U?Int(8,16,32,64)
+//
+
+// TODO use a more efficient encoder than `binary` (they say to in their top source comment)
+
+func EncodeNumber[T int8 | int16 | int32 | int64 | uint8 | uint16 | uint32 | uint64](w io.Writer, i T) (err error) {
+	return binary.Write(w, binary.BigEndian, i)
+}
+
+func DecodeNumber[T int8 | int16 | int32 | int64 | uint8 | uint16 | uint32 | uint64](r io.Reader) (i T, err error) {
+	err = binary.Read(r, binary.BigEndian, &i)
+	return
+}
+
+//
+// BigInt
+//
+
+func EncodeBigInt(w io.Writer, i *big.Int) (err error) {
+	isNegative := i.Sign() == -1
+	err = EncodeBool(w, isNegative)
+	if err != nil {
+		return
+	}
+
+	return EncodeBytes(w, i.Bytes())
```

Comment:
this does a copy of the underlying array, you could use `.Bits()` to get a little-endian word slice and iterate backwards.